### ISSUE TITLE
Return a promise from this.signals.mySignal()?

### ISSUE THREAD

**Issue by marbemac** [christianalfoni/cerebral#30]
This would be useful in a variety of circumstances. For example, when deleting a project from a react component, I'd like to route away (using the react-router context present in the component) if this.signals.deleteProject() succeeds. I also want to call another signal this.signals.createAppNotification('Project Deleted!') on success. 

```js
let ProjectPage = React.createClass({
  mixins: [ CerebralMixin ],

  getCerebralState() {
    return {
      project: ['currentProject']
    }
  },

  contextTypes: {
    router: React.PropTypes.func.isRequired
  },

  handleDelete() {
    let _this = this
    let r = confirm('Are you sure you want to delete this project? This will also delete all associated environments and requests.')
    if (r === true) {
      this.signals.deleteProject({id: this.state.project.id}).then(function() {
        _this.signals.createAppNotification('Project Deleted!')
        _this.context.router.transitionTo('entries')
      })
    }
  }

})
```

Does this make sense?

**Comment by christianalfoni** [comment]
Hmmm... this breaks the "one way flow" I think. It feels a bit hacky ;-)

This is one of the reasons why I think the react-router is not a good implementation. I think URLs should be state like anything else. So when you succeed, you change your "url-state" in an action, which a "ROUTER-COMPONENT", or some module, listens to and actually changes the url in the browser.

The app should be "state-driven", the react-router breaks that principle. The url and interface should change because of a state change, not because of a route change ;-)

I have actually been thinking about this quite a bit, let me go more in depth with this on next release.

**Comment by christianalfoni** [comment]
Okay, one way flow important for signals or Cerebral looses control :)

**Comment by pstoica** [comment]
Hey, sorry to bring this up again. How would you at least create this kind of ad hoc state in cerebral if you can't return a Promise? DELETE calls could probably add `$isDeleting` or `$isDeleted` flags, but what if you have something more local to a component?

Consider an object in cerebral state. If you want to update it, you need to make a PUT request, which will update the cerebral state upon success. It could be updated from several places without one component caring about another. It seems problematic to maintain global saving/error state. For example, validation errors should be unique to the components and shouldn't disappear unexpectedly.

Does cerebral have a way to create temporary state? You would at least be able to create side effects and Promises by listening for state changes.

**Comment by christianalfoni** [comment]
Hi @pstoica!

This is a really interesting topic indeed. I am sure that there are valid use cases where giving Cerebral control can feel like going uphill. But again, if you do not go through Cerebral you will not be able to scale in the sense that other components has access to that state. It will also be harder for you to reason about your code because there is no debugging information and the flow of those state changes work differently than all other state changes. But again, I absolutely see your point :-)

But this is how I think about it. When you define some state you define it because you need it in a component somewhere. So in practical terms using "getInitialState" inside a component or setting that exact same state inside Cerebral is no different. So a form inside a UserForm component could be:

```js
getInitialState() {
  return {
    isValid: true,
    isSubmitting: false,
    error: null,
    inputs: {
      name: '',
      age: ''
    }
  };
}

// And in cerebral
{
  userForm: {
    isValid: true,
    isSubmitting: false,
    error: null,
    inputs: {
      name: '',
      age: ''
    }
  }
}
```
So you just namespace it with something that relates to your application. And now you get some benefits:  

1. When for some reason some other component needs to know about this state, they can just grab it. For example locking other parts of the UI as the form is submitting
2. You can use signals to create the state change flow, which allows you to reuse existing actions
3. You can debug the information in the debugger
4. It works the same way as anything else

So yeah, you have to put your state somewhere and you have to create a state change flow to update that state. That does not change if its local or in Cerebral. But doing it in the Cerebral is more "future proof" :-)

But maybe we could create a more specific example and play around with it. We could write some code and see how different it would be?

**Comment by pstoica** [comment]
Hi @christianalfoni, thanks for the helpful response!

Namespacing some state makes a lot of sense. I didn't realize you could also dynamically add to the top of the `state` map (at least using cerebral-immutable; I hope this is an intentional feature :) ). Not defining state all at once is also great for codesplitting.

I'm coming from some frustrations with Flux and I can already see a lot of benefit from reusing actions that can request mutations *and* access state. Signals help define the action flow predictably, and I'm splitting up logic much less arbitrarily/redundantly.

In Flux, a lot of actions end up aggregating other actions through promises. It sometimes feels like an antipattern, especially when stores either return cached data or call some actions. `dispatch` and `action creators` and action aggregation are harder to expain; signals describe atomic flow more naturally, especially with async.

Anyway, it seems like Cerebral works against me less and your advice works fine, thank you!

Back to the original post, I think `createAppNotification` and updating the URL could be actions in the `deleteProject` signal. `componentWillReceiveProps` might be enough for producing the `transitionTo` side effect if you're using react-router?

**Comment by christianalfoni** [comment]
@pstoica Ah, yes, you can add to the top level too. That is not very clear with examples etc. :-)

An alternative is to split state definitions into different files like:

```js
import adminState from './state/adminState.js';
import homeState from './state/homeState.js';
const state = {
  admin: adminState,
  home: homeState
};
```
Structuring a Cerebral application is still being explored. Especially defining signals and state in larger applications. So the example above is one way to handle large applications with lots of state, and this is an example with signals:

```js
import appMountedSignal from './signals/appMounted.js';
cerebral.signal('appMounted', ...appMountedSignal);

// appMounted.js
export default [
  action1,
  action2,
  [
    action3, {
       success: [action3],
       error: [action4]
    }
  ]
];
```

Yeah, I totally agree with you. It is really difficult to structure your code when everything starts with one function. When scaling that function you have a tendency to just "put it in there". But with signals it is not one function, you are kinda forced to split up your flow into many functions... which is a really good thing :-)

Transitions requires a bit of a different mindset with a "general change event" system. Usually you might trigger a specific event to do a specific transition, its like a "reflex". But now this transition is based on a change of state.

One example of this could be to refresh an iframe on the page. An event initially makes more sense because you just want it to happen once, and I am not 100% settled on how we have to do this now. So:

1. Button triggers `iframeRefreshClicked`
2. Set `refreshIframe` to true?
3. Component checks it `refreshIframe` has changed from false to true
4. onLoad of iframe, or immediately , trigger a new signal: `iframeRefreshed`
5. set `refreshIframe` to false

Cerebral is inspired by how our bodies work (signal to brain etc.) and our bodies has a "reflex" system, I am just not sure how to express that with code. Especially in a way it can be controlled, like state changes are controlled. So in the case of the iframe it should be possible for the button component to just communicate directly with the iframe component, like a reflex... hm hm hm. Have to think about this more :-) Dont know if that made any sense, hehe... but I hope I answered your question!

**Comment by christianalfoni** [comment]
You know... Cerebral might getting something like:

```js
controller.reflex('iframeRefreshClicked');

And in the component you could:

@Reflex({
  iframeRefreshClicked: 'refreshIframe'
})
Comp extends React.Component {
  refreshIframe() {
    this.refs.iframe.reload();
  }
}
```
What I would have to figure out is how to control this with remembering state. It is really hard. When you normally remember state you trigger one change when all state is rebuilt. But how would you "remember a reflex"? Should it trigger all reflexes when state is rebuilt? Should it just trigger one of them?

A reflex has to be stateless though... that is for sure, and by nature they are. But again, hm... needs more thought :-)

**Comment by idream3** [comment]
@marbemac @christianalfoni 

Going back to the first comment, I suspect something like this would be possible... without testing though.

```javascript
// component.js
handleDelete() {
  const msg = 'Are you sure you want to delete this project? This will also delete all associated environments and requests.';
  
  if (confirm(msg)) {
    this.signals.deleteProjectConfirmed({
      id: this.state.project.id,
      router: this.context.router
    })
      
  }
}

// main.js
Controller.signal('deleteProjectConfirmed', 
  removeProjectOptimistic, // send this project as output to next actions
  [
    deleteProject: { // async to the server
      success: [
        showNotification('Project Deleted!'),  // factory method 
        transitionTo('entries') // uses the input.router and a factory method
      ],
      error: [
        rollbackProject, // using input.project from first action as input
        setErrorMessage
      ]
    }
  ]
)
```

@christianalfoni interesting idea about `reflex`...

**Comment by christianalfoni** [comment]
hi @idream3 ,

Interesting suggestion! It will not work due to inputs and inputs has to be JSON serializable, but everything else looks sweet :-)

But yeah, this depends on if we are talking about transitioning in general or transitioning routes. When it comes to routing it is better to use something like `reactive-router`, https://github.com/christianalfoni/reactive-router, which lets you control the router using state.

So basically in the component handling route transitions you would:

1. Check if the `currentPage` state has changed (should transition)
2. One container is "transitionOut" and one container is "transitionIn"
3. You render the previous currentPage in one and currentPage in the other
4. Run CSS transition or whatever

So yeah, hm... you know. This got me thinking. Looking at the iframe state again. A transition can not be handled with true/false, because a new transition can only happen if the state returns to false. But a transition is really some state changing to something else. This is easy to understand with a page:

1. If page is `home`
2. It changes to `admin`
3. It is different, transition
4. Page changes to `admin` again
5. It is the same, no transition
6. Page changes to `home`
7. It is different, transition

So with the iFrame example it should rather be handled like:

1. Create a state `lastIframeRequest: 0`
2. On button click, increase request to 1
3. IframeComponent checks if state has changed and does its thing
4. Next click and it gets to 2
5. IframeComponent sees state has changed and does its thing

So the thing is that. A transition is a sequence, it is not switch. Does this make sense? :-)

**Comment by christianalfoni** [comment]
You know what, my statement is that transitions should not be true/false, but arrays. Let me explain:

```js
let state = {
  someTransitions: []
};

const startTransition = function () {
  state.someTransitions.push(true);
};

const stopTransition = function () {
  state.someTransitions[state.someTransitions.length - 1] = false;
};

class TransitionComponent extends React.Component {
  componentWillUpdate(prevProps, prevState) {
    if (prevProps.someTransitions.length !== this.props.someTransitions.length) {
      // Run transition
    } else if (
      prevProps.someTransitions[prevProps.someTransitions.length - 1] !==
      this.props.someTransitions[prevProps.someTransitions.length - 1]
    ) {
      // Stop current transition
    }
  }
}
```
So transitions are a sequence of changes, where each step in sequence could hold information about current transition.

**Comment by bfitch** [comment]
@christianalfoni I like where you're heading with `controller.reflex('iframeRefreshClicked');`

That was my idea for component-to-component communication as well. Pretty much an Event Bus. I've used a similar pattern in plain React apps when I needed deeply nested child and parent components to communicate. 

```js
class Parent extends React.Component {
  componentWillMount() {
    EventBus.on("sidebarToggled", this._toggleSidebar, this);
  }
}

class Child extends React.Component {
  _toggleClicked(e)  {
    EventBus.trigger("sidebarToggled");
  }
}
```

Kinda a dumb example, but you get the idea. :smile: 

One concern is that introducing `reflex`-- state changes without going through `signals`-- complicates the clear mental model cerebral currently enables.  We would now have hands and feet that could communicate with each other without going through the brain!

I don't necessarily think `controller.events()` is *bad*, but it definitely would compromise the purity of the current design. Hmmmm....who knows!

**Comment by marbemac** [comment]
I'm not sure I understand why one would move away from state for this, and introduce "reflexes".

```js
//
// pseudo code below
//

// Initial state tree
var state = {
  sidebarOn: false
}

// Add some signals
controller.signals.sidebarToggled(function(args, state) {
  state.set('sidebarOn', !state.get('sidebarOn'))
})

class Parent extends React.Component {
  // use something like baobab or cerebral mixins to subscribe to state
  cursors: {
    sidebarOn: ['sidebarOn']  
  },

  render() {
    var sidebar
    if (this.state.sidebarOn) {
        sidebar = <div>Sidebar!</div>
    }
    return <div>{sidebar} foo</div>
  }
}

class Child extends React.Component {
  _toggleClicked(e)  {
    controller.signals.sidebarToggled()
  }
}
```

**Comment by bfitch** [comment]
Yeah @marbemac I agree that's how I would do it in a cerebral/baobab world too. There's probably a better example, but that was the first thing that popped into my head. 

However, just like there are some use cases for encapsulated component state (`setState`), perhaps there are some for direct component to component communications as well. Maybe?

I'm actually more interested in pursuing the current cerebral design of *all* state living in the tree, even though I was kinda advocating for other use cases just now. At a minimum, it is interesting to think about. :smile:

**Comment by christianalfoni** [comment]
Hi @marbemac ,

Ah, yeah, it is not a replacement. It would be an addition. Your example is state driven, but you have interactions in your app that is more easily triggered without state. An example of that would be if one component wants an other component to refresh an iframe. Its not an "on/off" state, its more a trigger. If that makes sense :-)

@bfitch Yeah, I agree it messes up a clear mental model. It would be a lot better to document use cases of how to solve these triggers/transitions or what you wanna call it. It least until a really good idea comes along :-)

**Comment by marbemac** [comment]
Yeah I thiiink I can understand that use case. Could we describe a real world example? What would cause component A to want to trigger an iframe update in component B, without any state change in the app, and without it being component b's responsibility?

There are so many benefits to managing as much state as possible in the canonical tree, outside of components. I worry introducing reflexes as described might confuse newcomers into keeping state in their components and using reflexes to start adding component - component communication. 

Basically, the current cerebral implementation embraces and encourages the declarative React way of things, and in my mind reflexes confuse that a bit. 

That said, if there are some real use cases where state doesn't work, I'm all for it! Just need to reduce possible confusion as much as possible.

**Comment by christianalfoni** [comment]
Yeah, I completely agree :-) I will do everything I can to avoid changing the simplicity that we have now, but I actually have a project where a button needs to refresh an iframe. Currently I actually do trigger an event, but as I mentioned above, the only way to really handle it with state is to have a counter.

So:

```js
class Parent extends React.Component {
  // use something like baobab or cerebral mixins to subscribe to state
  cursors: {
    iframRefreshRequests: ['iframRefreshRequests']  
  },
  componentDidUpdate(prevProps) {
    if (prevProps.iframRefreshRequests !== this.props.iframRefreshRequests) {
      this.refs.iframe.reload();
    }
  }
  render() {
    return <iframe ref="iframe"/>;
  }
}
```
So every time the button is clicked it triggers a signal that will increase a counter. That way our Parent component knows when it should refresh.

**Comment by pstoica** [comment]
I also had a use case for refreshing an iFrame! I toggled between true/false as mentioned above, but a counter's also a good solution.

Regarding reflexes, I kinda think it'd be useful to localize signals and state. I prototyped combining https://github.com/gaearon/react-stateful with local state that's backed in Cerebral state. The component would add/remove an entry from the `state.get('componentStates')` map, using signals for `componentWillMount`, `componentWillUnmount`, and `setState`. You end up with setState-like behavior + time traveling + debuggability almost for free.

Flux-style actions + reducer functions might be enough for localized state, but it could be interesting to incorporate the rest of Cerebral's signal and state APIs as a mini-cerebral. Building reflexes on top of this could also work well.

I did run into a problem when rewinding and causing an unmount, which triggers a signal. Is there a way to know if the user is seeking with the devtools, so you can change what happens in willMount/willUnmount?

Managing signals/state in the main tree still might be better and more predictable. I've been curious about reusing app-level state management at a component/local level, but I don't know if that's a bad thing in practice.

**Comment by christianalfoni** [comment]
Hi @pstoica ,

This sounds really interesting. At our office we were also playing around with the idea of defining the state inside the components, though they became global.

Cerebral should only trigger `change` when running normally. When remembering it triggers a `remember` event. It does not trigger any events when rebuilding state. You can also point into `controller.store.isRemembering()` to check.

The problem with local state in components is that it is more difficult to get an overview of the state in your app. Easier to implement, but harder to scale I think. And that is a challenge in general. Too often we do the "easy" thing... but that is just a statement, it would be great to investigate this further!
